### PR TITLE
add keepalive

### PR DESCRIPTION
**Pull request by day253**
fix
https://github.com/kitex-contrib/registry-zookeeper/issues/10

### CODE REVIEW

**Comment by horizonzy** [comment]
Could we use a watcher to listen to connection events? It's more graceful.

**Comment by horizonzy** [comment]
> Could we use a watcher to listen to connection events? It's more graceful.
Ignore it, the zk lib didn't support connection listener

File: registry/registry.go [hunk 1]
```diff
@@ -150,3 +153,48 @@ func (z *zookeeperRegistry) deleteNode(path string) error {
...
+
+func (z *zookeeperRegistry) ensureName(path string, data []byte, flags int32) error {
+	if z.conn == nil {
+		return errors.New("zk connection closed")
+	}
+	exists, stat, err := z.conn.Exists(path)
+	if err != nil {
+		return err
+	}
+	if flags&zk.FlagEphemeral == zk.FlagEphemeral {
```
Comment (horizonzy) [inline registry/registry.go]: If the `exists` is `false`, needn't delete it.

File: registry/registry.go [hunk 2]
```diff
@@ -150,3 +153,48 @@ func (z *zookeeperRegistry) deleteNode(path string) error {
 	}
 	return nil
 }
+
+func (z *zookeeperRegistry) keepalive(path string, content []byte) {
+	sessionID := z.conn.SessionID()
+	ticker := time.NewTicker(time.Second)
```
Comment (horizonzy) [inline registry/registry.go]: Second is too frequent; we can use the `sessionTimeout / 2`.

File: registry/registry.go [hunk 3]
```diff
@@ -150,3 +154,48 @@ func (z *zookeeperRegistry) deleteNode(path string) error {
...
+	sessionID := z.conn.SessionID()
+	ticker := time.NewTicker(z.sessionTimeout / 2)
+	defer ticker.Stop()
+	for range ticker.C {
+		cur := z.conn.SessionID()
+		// sessionID changed
+		if cur > 0 && sessionID != cur {
+			// re-ensureName
+			if err := z.ensureName(path, content, zk.FlagEphemeral); err != nil {
+				return
```
Comment (horizonzy) [inline registry/registry.go]: If we return, does the `keepalive` still work? If the zk server shutdown, it will return err.

Comment (day253) [reply registry/registry.go]: I check th zk lib. We should disable keepalive only when the connection is closed. We should alway try to keep the server registered when the zk servers shutdown.

Comment (day253) [reply registry/registry.go]: failed keepalive only when ErrConnectionClosed && only keepalive when Register success

File: registry/registry.go [hunk 4]
```diff
@@ -35,19 +35,33 @@ type zookeeperRegistry struct {
 	sessionTimeout time.Duration
 }
 
+var (
+	ErrConnectedTimedOut = errors.New("zk: connected timeout")
+)
+
 func NewZookeeperRegistry(servers []string, sessionTimeout time.Duration) (registry.Registry, error) {
```
Comment (day253) [inline registry/registry.go:44]: If the zk servers shutdown the server will wait until timeout to start.

File: registry/registry.go (hunk 3 above)
Comment (horizonzy) [reply registry/registry.go]: If the zkServer shutdown and restart, the keepalive won't work anymore.

Comment (day253) [reply registry/registry.go]: We should reconstruct the connection when the zk.ErrConnectionClosed happens?

Comment (day253) [reply registry/registry.go]: When zk.ErrConnectionClosed happens we cannot do anything use z.conn

https://github.com/go-zookeeper/zk/blob/master/conn.go#L429

If z.conn.Close is called.

Comment (horizonzy) [reply registry/registry.go]: I mean that the zkServer is down, not our registry client. 
My thought: we maintain the service info local, and listen to the connection change event, when received the connection change event, to check if the service is still registered, if not registered, re-register.
But now, the zk lib didn't support listening to connection change events. Using a timer to check the session change or not is also a good way. We needn't cancel the timer task, there is one timer task to check the connection is changed is enough.

File: registry/registry.go (hunk 4 above)
Comment (horizonzy) [reply registry/registry.go:44]: The zkServer may shutdown after registering.

File: registry/registry.go (hunk 3 above)
Comment (day253) [reply registry/registry.go]: ok. Please review the final change.

File: registry/registry.go [hunk 5]
```diff
@@ -150,3 +153,42 @@ func (z *zookeeperRegistry) deleteNode(path string) error {
...
+		}
+	}
+}
+
+func (z *zookeeperRegistry) ensureName(path string, data []byte, flags int32) error {
+	exists, stat, err := z.conn.Exists(path)
+	if err != nil {
+		return err
+	}
+	if exists && flags&zk.FlagEphemeral == zk.FlagEphemeral {
```
Comment (horizonzy) [inline registry/registry.go]: If the node still exists, return directly. no need delete and create/

Comment (day253) [reply registry/registry.go]: I want to fix a race condition if the server carshes after it creates the node without using CreateProtectedEphemeralSequential()

https://github.com/go-zookeeper/zk/blob/v1.0.3/conn.go#L1102

Comment (day253) [reply registry/registry.go]: Maybe I have to add the same logic when call createNode()

Comment (day253) [reply registry/registry.go]: Fixed.

**Comment by horizonzy** [comment]
Nicely done, could you add a test for cover this case?

**Comment by day253** [comment]
> Nicely done, could you add a test for cover this case?

Ok.

**Comment by GuangmingLuo** [comment]
> [CLA assistant check](https://cla-assistant.io/kitex-contrib/registry-zookeeper?pullRequest=14) Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/kitex-contrib/registry-zookeeper?pullRequest=14) before we can accept your contribution.You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/kitex-contrib/registry-zookeeper?pullRequest=14) it.

@day253  sign CLA please.

**Comment by GuangmingLuo** [comment]
@horizonzy please confirm and handle this PR.

**Comment by day253** [comment]
We deployed this to our production environment. But the difference is when Deregister is called we will cancel the keepalive goroutine. @horizonzy Should we add this function?

**Comment by day253** [comment]
> Nicely done, could you add a test for cover this case?

If we have the test to test the keepalive function. We should have two method. The first one is to delete the registered node. The other one is delete the registered node and cancel the keepalive function. But now the Deregister is almost the same as deleteNode.

File: registry/registry.go [hunk 6]
```diff
@@ -150,3 +153,37 @@ func (z *zookeeperRegistry) deleteNode(path string) error {
...
+		cur := z.conn.SessionID()
+		if cur != 0 && sessionID != cur {
+			if err := z.ensureName(path, content, zk.FlagEphemeral); err == nil {
+				sessionID = cur
+			}
+		}
+	}
+}
+
+func (z *zookeeperRegistry) ensureName(path string, data []byte, flags int32) error {
```
Comment (bodhisatan) [inline registry/registry.go]: The function's name doesn't express what it wants to ensure. Maybe a different name would be better.

**Comment by horizonzy** [comment]
Sorry for the delay reply, I'm watching now.

**Comment by horizonzy** [comment]
> We deployed this to our production environment. But the difference is when Deregister is called we will cancel the keepalive goroutine. @horizonzy Should we add this function?

Yes, if the service already be deregister, we shouldn't make it keep-alive anymore.

File: registry/registry.go (hunk 6 above)
Comment (horizonzy) [reply registry/registry.go]: +1

**Comment by horizonzy** [comment]
> If we have the test to test the keepalive function. We should have two method. The first one is to delete the registered node. The other one is delete the registered node and cancel the keepalive function. But now the Deregister is almost the same as deleteNode.

We can delete the node path using zk directly, not using the registry.go function.

**Comment by day253** [comment]
@horizonzy I have fixed all the comments.

**Comment by day253** [comment]
cc @bodhisatan

**Comment by bodhisatan** [comment]
> cc @bodhisatan

get, i'm reviewing it

**Comment by horizonzy** [comment]
Sorry, I haven't used Go for over a year, I am worried that I might not have the ability to review. Please help to review it, thanks! @bodhisatan

**Comment by day253** [comment]
@bodhisatan any idea?

**Review by bodhisatan — APPROVED** [review]

**Comment by bodhisatan** [comment]
> @bodhisatan any idea?

LGTM

**Comment by day253** [comment]
> > @bodhisatan any idea?
> 
> 
> 
> LGTM

thx

**Pull request merged**